### ISSUE TITLE
user defined eval metric function has different data interface

### ISSUE THREAD

**Issue by zhangfiona** [Microsoft/LightGBM#1453]
Parameter datatypes are different when user defined eval_metric function is used in train function of training api and fit function of scikit-learn api. And the sample code is as follows:

```
# self-defined eval metric
# f(preds: array, train_data: Dataset) -> name: string, value: array, is_higher_better: bool
# binary error
def eval_metric(preds, train_data):
    labels = train_data.get_label()
    return 'error', np.mean(labels != (preds > 0.5)), False

# Self-defined eval_metric is used in fit function of scikit-learn api, which causes an error
clf = lgb.LGBMClassifier()
clf.fit(X, y, eval_set=[[X,y]], eval_metric=eval_metric, early_stopping_rounds=5)

# Self-defined eval_metric is used in train function of training api, which executes sucessfully
data = lgb.Dataset(data=X, label=y)
lgb.train(params=params, train_set=data, feval=eval_metric, early_stopping_rounds=5, valid_sets=[data])
 ```
Self-defined eval_metric function is _not universal_ in these two train methods. The problem exists in **_eval_function_wrapper(func) in sklearn.py**, and this decorator changes the parameter datatypes. 
The code in sklean.py is here: [https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/sklearn.py#L113](https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/sklearn.py#L113)

**Comment by guolinke** [comment]
This design is from xgboost, to make sure the smooth immigration.

**Comment by tobycheese** [comment]
Why does your eval_metric take the parameters in the order (preds, train_data)? The docs state the opposite order:

> Custom eval function expects a callable with following functions: func(y_true, y_pred)

_eval_function_wrapper(func) seems to swap the arguments, I think your way is correct and the docs are wrong. This is confusing...

**Comment by zhangfiona** [comment]
The two self defined eval functions are both correct, but apply to different python apis. In [training api](https://lightgbm.readthedocs.io/en/latest/Python-API.html#training-api), we should define eval metric like: eval_metric(preds, train_data), just as the example [https://github.com/Microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py#L149](https://github.com/Microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py#L149); Whereas in [scikit-learn api](https://lightgbm.readthedocs.io/en/latest/Python-API.html#scikit-learn-api), the eval metric should be defined like: func(y_true, y_pred) or other formats according to the number of parameters. I think it is confusing and inconvnient either.

I also check the self defined function in [xgboost](https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/sklearn.py), the self defined objective function defined in the same way, that objective decorator swaps the arguments, but the self defined eval function doesn't consider different numbers of parameters, so it maintain the consistency in the two python apis (training api and scikit-learn api) without any decorator.

**Comment by tobycheese** [comment]
Ah ok, thanks for this extensive explanation! So not only the type of the parameters differ between the APIs, but also their order. Most unfortunate. And in addition, for the training api, the functionality is not documented except in the example (or at least I couldn't find it).

Thanks!

**Comment by guolinke** [comment]
@StrikerRUS can you confirm that ?

**Comment by zhangfiona** [comment]
@tobycheese , for training api, the functionality is documented here: https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L41

**Comment by StrikerRUS** [comment]
@zhangfiona Many thanks for the detailed explanation.

@guolinke 
>can you confirm that ?

What exactly?

I've just checked the docs and everything seems to be correct (just like @zhangfiona described).

@tobycheese Yes, signatures of eval functions in `train`/`cv` differ from those in sklearn wrapper.
